### PR TITLE
Use lazy initializers in place of double check locks

### PR DESCRIPTION
**Pull request by WinCPP**
Next set of changes for Issue 3862 to use lazy initialization constructs in place of double check locks.

@stephentoub @karelz kindly review.

### CODE REVIEW

File: src/Common/src/System/LocalAppContext.cs [hunk 1]
```diff
@@ -3,14 +3,15 @@
...
+using System.Threading;
 
 namespace System
 {
     internal partial class LocalAppContext
     {
         private static bool s_isDisableCachingInitialized;
         private static bool s_disableCaching;
-        private static readonly object s_syncObject = new object();
+        private static object s_syncObject = new object();
```
Comment (stephentoub) [inline src/Common/src/System/LocalAppContext.cs]: You can remove the `= new object()` and let it be lazily initialized.

File: src/Common/src/System/Net/SafeCloseSocket.Unix.cs [hunk 2]
```diff
@@ -98,18 +98,7 @@ public void TrackOption(SocketOptionLevel level, SocketOptionName name)
...
-            {
-                if (Volatile.Read(ref _asyncContext) == null)
-                {
-                    Interlocked.CompareExchange(ref _asyncContext, new SocketAsyncContext(this), null);
-                }
-
-                return _asyncContext;
-            }
-        }
+        public SocketAsyncContext AsyncContext => LazyInitializer.EnsureInitialized(ref _asyncContext, () => new SocketAsyncContext(this));
```
Comment (stephentoub) [inline src/Common/src/System/Net/SafeCloseSocket.Unix.cs]: Undo this.  This is adding a delegate allocation on every property access.

File: src/Common/src/System/Net/SafeCloseSocket.Windows.cs [hunk 3]
```diff
@@ -39,44 +33,37 @@ public ThreadPoolBoundHandle GetOrAllocateThreadPoolBoundHandle(bool trySkipComp
                 throw new ObjectDisposedException(typeof(Socket).FullName);
             }
 
-            // Check to see if the socket native _handle is already
-            // bound to the ThreadPool's completion port.
-            if (_iocpBoundHandle == null)
+            // Lazily bind the socket native _handle to the ThreadPool's completion port.
+            LazyInitializer.EnsureInitialized(ref _iocpBoundHandle, ref _iocpBindingLock, () =>
```
Comment (stephentoub) [inline src/Common/src/System/Net/SafeCloseSocket.Windows.cs]: Undo this.  This is adding a delegate allocation and closure allocation on every call.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs [hunk 4]
```diff
@@ -22,11 +23,11 @@ public sealed class LicenseManager
     {
         private static readonly object s_selfLock = new object();
 
-        private static volatile LicenseContext s_context;
+        private static LicenseContext s_context;
         private static object s_contextLockHolder;
         private static volatile Hashtable s_providers;
         private static volatile Hashtable s_providerInstances;
-        private static readonly object s_internalSyncObject = new object();
+        private static object s_internalSyncObject = new object();
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs]: You can remove the `= new object()` as well.

Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs]: Actually, no, I take that back, it's used elsewhere and needs to be initialized.  In which case it's also better for it to be readonly.  So let's just revert the changes here.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs [hunk 5]
```diff
@@ -599,18 +564,9 @@ public CLRLicenseContext(LicenseUsageMode usageMode, Type type)
...
-                    return type == _type ? _key : null;
-                }
-                public override void SetSavedLicenseKey(Type type, string key)
-                {
-                    if (type == _type)
-                    {
-                        _key = key;
-                    }
-                }
+                public override void SetSavedLicenseKey(Type type, string key) => _key = (type == _type) ? key : _key;
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs]: Why are you assigning _key to itself?

File: src/System.Console/src/System/Console.cs [hunk 6]
```diff
@@ -25,33 +25,16 @@ public static class Console
...
-                T result = Volatile.Read(ref field);
-                if (result == null)
-                {
-                    result = initializer();
-                    Volatile.Write(ref field, result);
-                }
-                return result;
-            }
-        }
+        internal static T EnsureInitializedWrapper<T>(ref T field, Func<T> initializer) where T : class =>
```
Comment (stephentoub) [inline src/System.Console/src/System/Console.cs]: Why did you rename this?

File: src/System.Console/src/System/Console.cs [hunk 7]
```diff
@@ -25,33 +25,16 @@ public static class Console
...
+            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, () => initializer());
 
-        public static TextReader In
-        {
-            get
-            {
-                return Volatile.Read(ref s_in) ?? EnsureInitialized(ref s_in, () => ConsolePal.GetOrCreateReader());
-            }
-        }
+        public static TextReader In => EnsureInitializedWrapper(ref s_in, () => ConsolePal.GetOrCreateReader());
```
Comment (stephentoub) [inline src/System.Console/src/System/Console.cs]: You don't actually need EnsureInitialized; you could just delete that and instead use LazyInitializer.EnsureInitialized here directly:
```C#
public static TextReader In =>
    LazyInitializer.EnsureInitialized(ref s_in, ref InternalSyncObject, () => ConsolePal.GetOrCreateReader();
```
The extra generic method doesn't really help anything here.

File: src/System.Diagnostics.Process/src/System/Diagnostics/PerformanceCounterLib.cs [hunk 8]
```diff
@@ -45,41 +33,9 @@ internal PerformanceCounterLib(string machineName, string lcid)
...
-                    {
-                        if (_nameTable == null)
-                            _nameTable = GetStringTable(false);
-                    }
-                }
-
-                return _nameTable;
-            }
-        }
+        internal Dictionary<int, string> NameTable => LazyInitializer.EnsureInitialized(ref _nameTable, ref _nameTableLock, () => GetStringTable(false));
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/PerformanceCounterLib.cs]: Undo this.  This will allocate a delegate on every call.

**Comment by stephentoub** [comment]
Thanks for the efforts on this, but a bunch of the changes are adding non-trivial cost and need to be undone.  Please also undo all of the formatting changes; they're unrelated to the purpose of the PR and make it harder to review.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs (hunk 5 above)
Comment (WinCPP) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs]: Else it would not allow me to use ? : ternary operator... but I'm removing it as formatting changes unrelated to the PR are to be removed.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs (hunk 4 above)
Comment (WinCPP) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs]: I had to remove volatile at line 26 as it loses meaning while passing as ref and compiler gives error on it. In that flow, I wrongly removed the readonly on this field... Reverting.

File: src/Common/src/System/LocalAppContext.cs [hunk 9]
```diff
@@ -35,29 +36,12 @@ private static bool GetCachedSwitchValueInternal(string switchName, ref int swit
             return isSwitchEnabled;
         }
 
-        private static bool DisableCaching 
-        { 
-            get 
+        private static bool DisableCaching =>
+            LazyInitializer.EnsureInitialized(ref s_disableCaching, ref s_isDisableCachingInitialized, ref s_syncObject, () =>
```
Comment (WinCPP) [inline src/Common/src/System/LocalAppContext.cs]: This is for my learning. As with that at line 101, will this property access also allocate a delegate on each invocation? I wrote some test code for a static property and inspected it using ILSpy which generated following code,
```C#
private static HasDefaultCtor staticProperty
{
	get
	{
		Func<HasDefaultCtor> arg_2E_3;
		if ((arg_2E_3 = Program.<>c.<>9__6_0) == null)
		{
			arg_2E_3 = (Program.<>c.<>9__6_0 = new Func<HasDefaultCtor>(Program.<>c.<>9.<get_staticProperty>b__6_0));
		}
		return LazyInitializer.EnsureInitialized<HasDefaultCtor>(ref Program.s_defaultCtorObj, ref Program.s_hdcInited, ref Program.s_hdcLock, arg_2E_3);
	}
}
```
Looks like `arg_2E_3` will be allocated on each access for the static property as well? I am trying to understand how this is different from line 101 which also involves similar function call. Then would usage of LazyInitializer.EnsureInitialize with valueFactory always involve a delegate allocation, perhaps in those places (as this one) which was a plain function call previously? Please help me understand.

Comment (stephentoub) [inline src/Common/src/System/LocalAppContext.cs]: Any time a lambda/delegate closes over instance state, including implicitly over `this`, you'll get at least one allocation for the delegate, and if more than `this` is needed, you'll also get a closure object to store all of the state.  If the lambda/delegate only ever accesses static state, then the compiler can avoid any closure allocation and can generate code that lazily-initializes a delegate for the call site, caching it for subsequent access.

Comment (WinCPP) [inline src/Common/src/System/LocalAppContext.cs]: @stephentoub sorry I updated the comment. Kindly check once again. This time I'm trying to understand instance property...

Comment (WinCPP) [inline src/Common/src/System/LocalAppContext.cs]: Okk got it... the crux is instance state. Understood with this sample,

```C#
// StackObjectCheck.Program
private HasDefaultCtor instanceProperty
{
	get
	{
		return LazyInitializer.EnsureInitialized<HasDefaultCtor>(ref this.instanceObject, ref Program.s_hdcInited, ref Program.s_hdcLock, () => new HasDefaultCtor(this));
	}
}
```

File: src/System.Console/src/System/Console.cs (hunk 6 above)
Comment (WinCPP) [inline src/System.Console/src/System/Console.cs]: I renamed it to not cause confusion with LazyInitializer.EnsureInitialized as I wanted to retain this method. In addition, w.r.t. comment below, I wanted to retain this method as it wraps the sync object shared via usages of this EnsureInitialized method. Eliminating this method would require spilling out the lock as well, with potential to use another lock; essentially making the code semantically unequal to what author had written...

File: src/System.Console/src/System/Console.cs (hunk 7 above)
Comment (WinCPP) [inline src/System.Console/src/System/Console.cs]: Please refer to above comment for why I retained the EnsureInitialized method, although with rename.

File: src/System.Console/src/System/Console.cs (hunk 6 above)
Comment (WinCPP) [inline src/System.Console/src/System/Console.cs]: Is that ok? Because otherwise, I will have to spill out InternalSyncObject in multiple places where EnsureInitialized(Wrapper) has been used.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs]: Sorry, when I wrote "let's just revert the changes here", I meant in the whole file.  There are many other uses of s_internalSyncObject throughout the file, and I would rather we have the safety of that being readonly that save a few lines of code by using EnsureInitialized.

File: src/System.Console/src/System/Console.cs (hunk 6 above)
Comment (stephentoub) [inline src/System.Console/src/System/Console.cs]: If we keep the method, I would prefer it just be named `EnsureInitialized`.  Callers of the method don't care how it's implemented and that it happens to be a wrapper for LazyInitialized.EnsureInitialized, all they care is that it's ensuring that the thing they need is initialized.

File: src/System.Console/src/System/Console.cs [hunk 10]
```diff
@@ -25,33 +25,16 @@ public static class Console
...
-                if (result == null)
-                {
-                    result = initializer();
-                    Volatile.Write(ref field, result);
-                }
-                return result;
-            }
-        }
+        internal static T EnsureInitializedWrapper<T>(ref T field, Func<T> initializer) where T : class =>
+            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, () => initializer());
```
Comment (stephentoub) [inline src/System.Console/src/System/Console.cs]: Can't the last argument just be `initializer` rather than `, () => initializer()`?

File: src/System.Console/src/System/Console.cs [hunk 11]
```diff
@@ -153,35 +130,11 @@ private static TextWriter CreateOutputWriter(Stream outputStream)
...
-        public static bool IsInputRedirected
-        {
-            get
-            {
-                StrongBox<bool> redirected = Volatile.Read(ref _isStdInRedirected) ??
-                    EnsureInitialized(ref _isStdInRedirected, () => new StrongBox<bool>(ConsolePal.IsInputRedirectedCore()));
-                return redirected.Value;
-            }
-        }
+        public static bool IsInputRedirected => EnsureInitializedWrapper(ref _isStdInRedirected, () => new StrongBox<bool>(ConsolePal.IsInputRedirectedCore())).Value;
```
Comment (stephentoub) [inline src/System.Console/src/System/Console.cs]: This is a very long line.  Please revert the formatting to be more like what it was. Same below.

File: src/System.Diagnostics.Process/src/System/Diagnostics/PerformanceCounterLib.cs [hunk 12]
```diff
@@ -116,15 +80,7 @@ internal static PerformanceCounterLib GetPerformanceCounterLib(string machineNam
...
-            if (_performanceMonitor == null)
-            {
-                lock (InternalSyncObject)
-                {
-                    if (_performanceMonitor == null)
-                        _performanceMonitor = new PerformanceMonitor(_machineName);
-                }
-            }
-
+            LazyInitializer.EnsureInitialized(ref _performanceMonitor, ref s_internalSyncObject, () => new PerformanceMonitor(_machineName));
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/PerformanceCounterLib.cs]: Revert this; the lambda is closing over instance state.

**Comment by stephentoub** [comment]
Thanks, @WinCPP.  I left some more comments for things to be addressed.

**Comment by WinCPP** [comment]
@stephentoub please check.

File: src/System.Console/src/System/Console.cs [hunk 13]
```diff
@@ -157,29 +134,26 @@ public static bool IsInputRedirected
         {
             get
             {
-                StrongBox<bool> redirected = Volatile.Read(ref _isStdInRedirected) ??
-                    EnsureInitialized(ref _isStdInRedirected, () => new StrongBox<bool>(ConsolePal.IsInputRedirectedCore()));
-                return redirected.Value;
+                EnsureInitialized(ref _isStdInRedirected, () => new StrongBox<bool>(ConsolePal.IsInputRedirectedCore()));
+                return _isStdInRedirected.Value;
```
Comment (stephentoub) [inline src/System.Console/src/System/Console.cs]: Nit: it'd be better to store the result of EnsureInitialized into a local, as the previous code had it, e.g.
```C#
var redirected = EnsureInitialized(...);
return redirect.Value;
```

File: src/System.Console/src/System/Console.cs [hunk 14]
```diff
@@ -157,29 +134,26 @@ public static bool IsInputRedirected
...
 
         public static bool IsOutputRedirected
         {
             get
             {
-                StrongBox<bool> redirected = Volatile.Read(ref _isStdOutRedirected) ??
-                    EnsureInitialized(ref _isStdOutRedirected, () => new StrongBox<bool>(ConsolePal.IsOutputRedirectedCore()));
-                return redirected.Value;
+                EnsureInitialized(ref _isStdOutRedirected, () => new StrongBox<bool>(ConsolePal.IsOutputRedirectedCore()));
+                return _isStdOutRedirected.Value;
```
Comment (stephentoub) [inline src/System.Console/src/System/Console.cs]: Ditto

File: src/System.Console/src/System/Console.cs [hunk 15]
```diff
@@ -157,29 +134,26 @@ public static bool IsInputRedirected
...
 
         public static bool IsErrorRedirected
         {
             get
             {
-                StrongBox<bool> redirected = Volatile.Read(ref _isStdErrRedirected) ??
-                    EnsureInitialized(ref _isStdErrRedirected, () => new StrongBox<bool>(ConsolePal.IsErrorRedirectedCore()));
-                return redirected.Value;
+                EnsureInitialized(ref _isStdErrRedirected, () => new StrongBox<bool>(ConsolePal.IsErrorRedirectedCore()));
+                return _isStdErrRedirected.Value;
```
Comment (stephentoub) [inline src/System.Console/src/System/Console.cs]: Ditto

File: src/System.Diagnostics.Process/src/System/Diagnostics/PerformanceCounterLib.cs [hunk 16]
```diff
@@ -118,7 +82,7 @@ internal byte[] GetPerformanceData(string item)
         {
             if (_performanceMonitor == null)
             {
-                lock (InternalSyncObject)
+                lock (s_internalSyncObject)
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/PerformanceCounterLib.cs]: Won't this fail with a null reference exception if s_internalSyncObject hasn't been initialized yet?  Or does something guarantee that it has been initialized?

Comment (WinCPP) [inline src/System.Diagnostics.Process/src/System/Diagnostics/PerformanceCounterLib.cs]: My bad that I missed the requirement for lazy initialization.

**Comment by stephentoub** [comment]
Thanks.
